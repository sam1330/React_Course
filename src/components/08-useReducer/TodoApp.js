import React, { useReducer, useEffect } from "react";

import { todoReducer } from "./todoReducer";

import "./styles.css";

import TodoList from "./components/TodoList";
import TodoAdd from "./components/TodoAdd";

const init = () => {

  return JSON.parse(localStorage.getItem('todos')) || [];

};

const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);


  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos])

  const handleDelete = (todoId) => {
    dispatch({
      type: "delete",
      payload: todoId
    });
  };


  const handleAddTodo = (newTodo) => {
    dispatch({
      type: "add",
      payload: newTodo
    });
  };

  const handleToggle = (todoId) => {
    dispatch({
      type: "toggle",
      payload: todoId
    });
  };

  return (
    <div className="container">
      <h1 className="font-weight-bold">Todo App ( {todos.length} )</h1>
      <hr />
      <div className="row">
        <div className="col-6">
          <h4 className="font-weight-bold">Todos</h4>
          <hr />
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-6">
          <TodoAdd handleAddTodo={ handleAddTodo } />
        </div>
      </div>
    </div>
  )
};

export default TodoApp;