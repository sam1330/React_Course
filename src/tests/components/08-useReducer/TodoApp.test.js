import { act } from "@testing-library/react";
import { shallow } from "enzyme";
import TodoApp from "../../../components/08-useReducer/TodoApp";
import { demoTodos } from "../../fixtures/demoTodos";


describe('<TodoApp /> tests', () => {

  const wrapper = shallow(<TodoApp />);

  Storage.prototype.setItem = jest.fn();

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should add todo', () => {
    const wrapper = shallow(<TodoApp />);

    act(() => {
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
    });

    expect(wrapper.find('TodoList').prop('todos').length).toBe(2);
  });

  test('should delete todo', () => {

    wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
    expect(wrapper.find('h1').text().trim()).toBe('Todo App ( 1 )');

    wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);
    expect(wrapper.find('h1').text().trim()).toBe('Todo App ( 0 )');

  });
});