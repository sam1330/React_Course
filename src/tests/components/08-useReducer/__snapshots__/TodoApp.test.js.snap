// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoApp /> tests should render the component 1`] = `
<div
  className="container"
>
  <h1
    className="font-weight-bold"
  >
    Todo App ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-6"
    >
      <h4
        className="font-weight-bold"
      >
        Todos
      </h4>
      <hr />
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-6"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
